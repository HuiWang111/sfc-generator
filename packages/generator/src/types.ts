import type generate from '@babel/generator'
import type { GeneratorOptions } from '@babel/generator'
import type { SFCBlock } from '@vue/compiler-sfc'
import type { TemplateNodeType } from './constants'

export interface TemplateNodeAttr {
  name: string
  value: string
  dynamic?: boolean | undefined
}

export interface TemplateNodeDirective {
  name: string
  rawName: string
  value: string
  arg: null
  isDynamicArg: false
  modifiers: null
}

export interface TemplateNodeModel {
  value: string
  expression: string // eg: '"value"'
  callback: string
}

export type TemplateNode =
  | ElementNode
  | InterpolationNode
  | TextNode

export interface ElementNode {
  type: TemplateNodeType.Element
  tag: string
  attrs?: TemplateNodeAttr[]
  attrsList?: Omit<TemplateNodeAttr, 'dynamic'>[]
  attrsMap?: Record<string, string> // 包含 events / class
  rawAttrsMap?: Record<string, string>
  parent?: TemplateNode | undefined
  children: TemplateNode[]
  plain?: boolean
  static?: boolean
  staticRoot?: boolean
  events?: Record<string, {
    value: string
    dynamic: boolean
  }>
  staticClass?: string // eg: '"submit-container"'
  hasBindings?: boolean
  forProcessed?: boolean
  alias?: string // v-for 定义的变量，eg: v-for="data in list" 中的 data
  for?: string // 被循环的变量，eg: v-for="data in list" 中的 list
  directives?: TemplateNodeDirective[]
  model?: TemplateNodeModel
  __skip__?: boolean
}

export interface InterpolationNode {
  type: TemplateNodeType.Interpolation
  tokens: object[]
  expression: string
  text: string
  parent?: TemplateNode | undefined
  __skip__?: boolean
}

export interface TextNode {
  type: TemplateNodeType.Text
  text: string
  parent?: TemplateNode | undefined
  __skip__?: boolean
}

export interface GenerateOptions {
  breakOnAttrs?: number
  indentSize?: number
  autoIndent?: boolean
}

export type BabelNode = Parameters<typeof generate>[0]

export interface ScriptAttrs {
  setup?: boolean
  lang?: string
}

export interface SFCGenerateOptions {
  template: {
    node: TemplateNode
    options?: GenerateOptions
  }
  script?: {
    node: BabelNode
    attrs?: ScriptAttrs
    options?: GeneratorOptions
  }
  styles?: SFCBlock[]
}

export interface ScriptParseOptions<T = boolean> {
  setup?: T
  lang?: 'ts' | 'js'
  jsx?: boolean
}

export interface OptionOperator {
  add: (...args: any[]) => any
  update: (...args: any[]) => any
  remove: (name: string) => any
  get: (name: string) => any
}
