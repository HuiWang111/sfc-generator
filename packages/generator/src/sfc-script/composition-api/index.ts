import type { Statement } from '@babel/types'
import {
  ComputedOption,
  PropsOption,
  RefOption,
} from './properties'

export class CompositionApi {
  refOption: RefOption | null = null
  computedOption: ComputedOption | null = null
  propsOption: PropsOption | null = null

  constructor(private statements: Statement[]) {}

  data() {
    if (!this.refOption)
      this.refOption = new RefOption(this.statements)
    return this.refOption
  }

  computed() {
    if (!this.computedOption)
      this.computedOption = new ComputedOption(this.statements)
    return this.computedOption
  }

  props() {
    if (!this.propsOption)
      this.propsOption = new PropsOption(this.statements)
    return this.propsOption
  }
}
