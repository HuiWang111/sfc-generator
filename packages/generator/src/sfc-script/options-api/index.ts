import type { NodePath } from '@babel/traverse'
import type { ObjectExpression, ObjectMethod, ObjectProperty } from '@babel/types'
import * as t from '@babel/types'
import {
  ComputedOption,
  DataOption,
  MethodsOption,
  PropsOption,
  WatchOption,
} from './properties'

export class OptionsApi {
  // private isSetup: boolean
  private dataOption: DataOption | null = null
  private computedOption: ComputedOption | null = null
  private watchOption: WatchOption | null = null
  private methodsOption: MethodsOption | null = null
  private propsOption: PropsOption | null = null

  constructor(private _node: NodePath<ObjectExpression>) {}

  private getOptionNode<T extends ObjectMethod | ObjectProperty>(name: string) {
    return this._node.get('properties')
      .find((prop): prop is NodePath<T> => {
        const keyNode = prop.get('key')
        return keyNode.isIdentifier() && keyNode.node.name === name
      })
  }

  get node() {
    return this._node
  }

  data() {
    if (this.dataOption) {
      return this.dataOption
    }

    const dataNode = this.getOptionNode<ObjectMethod>('data')
    if (dataNode) {
      this.dataOption = new DataOption(dataNode.node, this)
      return this.dataOption
    }

    const objectMethod = t.objectMethod(
      'method',
      t.identifier('data'),
      [],
      t.blockStatement([
        t.returnStatement(
          t.objectExpression([]),
        ),
      ]),
    )
    this._node.pushContainer('properties', objectMethod)
    this.dataOption = new DataOption(objectMethod, this)
    return this.dataOption
  }

  computed() {
    if (this.computedOption)
      return this.computedOption

    const computedNode = this.getOptionNode<ObjectProperty>('computed')
    if (computedNode) {
      this.computedOption = new ComputedOption(computedNode.node, this)
      return this.computedOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('computed'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.computedOption = new ComputedOption(objectProperty, this)
    return this.computedOption
  }

  watch() {
    if (this.watchOption)
      return this.watchOption

    const watchNode = this.getOptionNode<ObjectProperty>('watch')
    if (watchNode) {
      this.watchOption = new WatchOption(watchNode.node, this)
      return this.watchOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('watch'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.watchOption = new WatchOption(objectProperty, this)
    return this.watchOption
  }

  methods() {
    if (this.methodsOption)
      return this.methodsOption

    const methodsNode = this.getOptionNode<ObjectProperty>('methods')
    if (methodsNode) {
      this.methodsOption = new MethodsOption(methodsNode.node, this)
      return this.methodsOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('methods'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.methodsOption = new MethodsOption(objectProperty, this)
    return this.methodsOption
  }

  props() {
    if (this.propsOption)
      return this.propsOption

    const propsNode = this.getOptionNode<ObjectProperty>('props')
    if (propsNode) {
      this.propsOption = new PropsOption(propsNode.node, this)
      return this.propsOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('props'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.propsOption = new PropsOption(objectProperty, this)
    return this.propsOption
  }
}
