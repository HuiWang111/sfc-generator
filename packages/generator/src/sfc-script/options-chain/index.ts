export { ObjectExpressionOptionsChain } from './object-expression'
