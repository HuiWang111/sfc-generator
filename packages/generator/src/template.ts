import type { ElementNode, GenerateOptions, TemplateNode } from './types'
import { TemplateNodeType } from './types'

export class Template {
  constructor(private options: GenerateOptions = {}) {}

  private padCode(str: string, indentCount: number) {
    const { indent = 2 } = this.options
    const spaceCount = indent * indentCount
    return str.padStart(str.length + spaceCount)
  }

  private generateAttrCode(
    node: ElementNode,
    indentCount: number,
  ): [string, boolean] {
    const { breakOnAttrs = 2 } = this.options
    const attrEntries = Object.entries(node.attrsMap)
    if (!attrEntries.length) {
      return ['', false]
    }

    const shouldBreak = attrEntries.length > breakOnAttrs
    const attrCodes = (
      attrEntries
        .map(([key, value]) => {
          if (shouldBreak) {
            return this.padCode(`${key}="${value}"`, indentCount)
          }
          return `${key}="${value}"`
        })
    )

    if (!shouldBreak) {
      return [attrCodes.join(' '), shouldBreak]
    }
    return [attrCodes.join('\n'), shouldBreak]
  }

  private generateEndTag(
    node: ElementNode,
    indentCount: number,
    shouldBreak: boolean,
  ) {
    if (!node.children || !node.children.length) {
      return this.padCode('/>', indentCount)
    }
    const childrenCode = (
      node.children
        .map((c) => {
          return this.generate(c, indentCount + 1)
        })
        .join('')
    )
    return [
      this.padCode('>', shouldBreak ? indentCount : 0),
      childrenCode,
      this.padCode(`</${node.tag}>`, indentCount),
    ].join('\n')
  }

  generate(node: TemplateNode, indentCount = 1): string {
    if (
      node.type === TemplateNodeType.Interpolation
      || node.type === TemplateNodeType.Text
    ) {
      if (node.type === TemplateNodeType.Text && node.text === ' ') {
        return '\n'
      }
      return this.padCode(node.text, indentCount)
    }

    const [attrCode, shouldBreak] = this.generateAttrCode(node, indentCount + 1)
    const startTag = this.padCode(`<${node.tag}`, indentCount)
    const endTag = this.generateEndTag(node, indentCount, shouldBreak)

    if (!shouldBreak) {
      return [
        startTag,
        attrCode ? ' ' : '',
        attrCode,
        endTag,
      ]
        .join('')
    }
    return [startTag, attrCode, endTag].join('\n')
  }
}
