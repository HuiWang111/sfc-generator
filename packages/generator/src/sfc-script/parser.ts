import type { ParserOptions } from '@babel/parser'
import type { ScriptParseOptions } from '../types'
import { parse as babelParse } from '@babel/parser'
import traverse from '@babel/traverse'
import { CompositionApi } from './composition-api'
import { OptionsApi } from './options-api'

export function parse<T extends boolean = false>(
  code: string,
  options: ScriptParseOptions<T> = {},
): {
    api: T extends true ? CompositionApi : OptionsApi
    ast: ReturnType<typeof babelParse>
  } {
  const { lang, jsx = false, setup = false } = options

  const plugins: ParserOptions['plugins'] = []
  if (lang === 'ts') {
    plugins.push('typescript')
  }
  if (jsx) {
    plugins.push('jsx')
  }

  const ast = babelParse(code, {
    sourceType: 'module',
    plugins,
  })

  let api: OptionsApi | CompositionApi | null = null

  if (setup) {
    api = new CompositionApi(ast.program.body)
  }
  else {
    traverse(ast, {
      ExportDefaultDeclaration(path) {
        if (setup)
          return

        const declarationPath = path.get('declaration')
        if (declarationPath.isObjectExpression()) {
          api = new OptionsApi(declarationPath)
        }
      },
    })
  }

  if (!api) {
    throw new Error(`script setup attr is ${setup}, parse code error: ${code}`)
  }

  return {
    api: api as any,
    ast,
  }
}
