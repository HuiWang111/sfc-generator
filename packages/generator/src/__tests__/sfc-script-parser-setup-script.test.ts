import generate from '@babel/generator'
import template from '@babel/template'
import * as t from '@babel/types'
import { parse } from '../sfc-script'

describe('sfc setup script parser should work', () => {
  it('ref crud should work', () => {
    const js = 'const checked = ref(\'\')'

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.data().add(
      'value',
      t.stringLiteral(''),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.data().update('checked', t.booleanLiteral(false));
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.data().remove('checked');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const valueNode = api.data().get('value')
    expect(valueNode).toMatchSnapshot()
  })

  it('computed crud should work', () => {
    const js = `const firstName = ref('Davide');
const lastName = ref('Li')
const fullName = computed(() => firstName.value + lastName.value)`

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.computed().add(
      'reverseName',
      template.expression('() => lastName.value + \' \' + firstName.value')(),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.computed().update(
      'fullName',
      template.expression('() => firstName.value + \' \' + lastName.value')(),
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.computed().remove('fullName');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const valueNode = api.computed().get('reverseName')
    expect(valueNode).toMatchSnapshot()
  })

  it('update defineProps callExpression should work', () => {
    const js = `defineProps({
      value: {
        type: String,
        default: '',
      },
    })`

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.props().add(
      t.objectProperty(
        t.identifier('checked'),
        t.identifier('Boolean'),
      ),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.props().update(
      t.objectProperty(
        t.identifier('value'),
        t.identifier('String'),
      ),
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.props().remove('value');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const valueNode = api.props().get('checked')
    expect(valueNode).toMatchSnapshot()
  })

  it('update defineProps variableDeclaration should work', () => {
    const js = `const props = defineProps({
      value: {
        type: String,
        default: '',
      },
    })`

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.props().add(
      t.objectProperty(
        t.identifier('checked'),
        t.identifier('Boolean'),
      ),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.props().update(
      t.objectProperty(
        t.identifier('value'),
        t.identifier('String'),
      ),
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.props().remove('value');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const valueNode = api.props().get('checked')
    expect(valueNode).toMatchSnapshot()
  })

  it('create defineProps should work', () => {
    const js = `const str = ref('')`
    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.props().add(
      t.objectProperty(
        t.identifier('checked'),
        t.identifier('Boolean'),
      ),
    )
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('methods crud should work', () => {
    const js = `const inputValue = ref('')
    const checked = ref(false)

    function handleInput(val) {
      inputValue.value = val
    }`

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.methods().add(
      template.statement(`function handleChange(val){
        checked.value = val
      }`)() as t.FunctionDeclaration,
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.methods().update(
      template.statement(`function handleInput(value) {
        inputValue.value = value
      }`)() as t.FunctionDeclaration,
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.methods().remove('handleInput');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const valueNode = api.methods().get('handleChange')
    expect(valueNode).toMatchSnapshot()
  })

  it('watch crud should work', () => {
    const js = `const x = ref(0)`

    const { ast, api } = parse<true>(js, { setup: true })
    if (!api)
      return

    api.watch().add(
      template.statement('watch(x, (newX) => {\nconsole.log(`x is ${newX}`)\n})')() as t.ExpressionStatement
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
