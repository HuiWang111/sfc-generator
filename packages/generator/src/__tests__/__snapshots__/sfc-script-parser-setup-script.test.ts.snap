// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sfc setup script parser should work computed crud should work 1`] = `
"const firstName = ref('Davide');
const lastName = ref('Li');
const fullName = computed(() => firstName.value + lastName.value);
const reverseName = computed(() => lastName.value + ' ' + firstName.value);"
`;

exports[`sfc setup script parser should work computed crud should work 2`] = `
"const firstName = ref('Davide');
const lastName = ref('Li');
const fullName = computed(() => firstName.value + ' ' + lastName.value);
const reverseName = computed(() => lastName.value + ' ' + firstName.value);"
`;

exports[`sfc setup script parser should work computed crud should work 3`] = `
"const firstName = ref('Davide');
const lastName = ref('Li');
const reverseName = computed(() => lastName.value + ' ' + firstName.value);"
`;

exports[`sfc setup script parser should work computed crud should work 4`] = `
{
  "async": false,
  "body": {
    "left": {
      "left": {
        "computed": false,
        "loc": undefined,
        "object": {
          "loc": undefined,
          "name": "lastName",
          "type": "Identifier",
        },
        "property": {
          "loc": undefined,
          "name": "value",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "loc": undefined,
      "operator": "+",
      "right": {
        "extra": {
          "raw": "' '",
          "rawValue": " ",
        },
        "loc": undefined,
        "type": "StringLiteral",
        "value": " ",
      },
      "type": "BinaryExpression",
    },
    "loc": undefined,
    "operator": "+",
    "right": {
      "computed": false,
      "loc": undefined,
      "object": {
        "loc": undefined,
        "name": "firstName",
        "type": "Identifier",
      },
      "property": {
        "loc": undefined,
        "name": "value",
        "type": "Identifier",
      },
      "type": "MemberExpression",
    },
    "type": "BinaryExpression",
  },
  "extra": {
    "parenStart": 0,
    "parenthesized": true,
  },
  "generator": false,
  "loc": undefined,
  "params": [],
  "type": "ArrowFunctionExpression",
}
`;

exports[`sfc setup script parser should work create defineProps should work 1`] = `
"defineProps({
  checked: Boolean
});
const str = ref('');"
`;

exports[`sfc setup script parser should work ref crud should work 1`] = `
"const checked = ref('');
const value = ref("");"
`;

exports[`sfc setup script parser should work ref crud should work 2`] = `
"const checked = ref(false);
const value = ref("");"
`;

exports[`sfc setup script parser should work ref crud should work 3`] = `"const value = ref("");"`;

exports[`sfc setup script parser should work ref crud should work 4`] = `
{
  "type": "StringLiteral",
  "value": "",
}
`;

exports[`sfc setup script parser should work update defineProps callExpression should work 1`] = `
"defineProps({
  value: {
    type: String,
    default: ''
  },
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps callExpression should work 2`] = `
"defineProps({
  value: String,
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps callExpression should work 3`] = `
"defineProps({
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps callExpression should work 4`] = `
{
  "computed": false,
  "decorators": null,
  "key": {
    "name": "checked",
    "type": "Identifier",
  },
  "shorthand": false,
  "type": "ObjectProperty",
  "value": {
    "name": "Boolean",
    "type": "Identifier",
  },
}
`;

exports[`sfc setup script parser should work update defineProps variableDeclaration should work 1`] = `
"const props = defineProps({
  value: {
    type: String,
    default: ''
  },
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps variableDeclaration should work 2`] = `
"const props = defineProps({
  value: String,
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps variableDeclaration should work 3`] = `
"const props = defineProps({
  checked: Boolean
});"
`;

exports[`sfc setup script parser should work update defineProps variableDeclaration should work 4`] = `
{
  "computed": false,
  "decorators": null,
  "key": {
    "name": "checked",
    "type": "Identifier",
  },
  "shorthand": false,
  "type": "ObjectProperty",
  "value": {
    "name": "Boolean",
    "type": "Identifier",
  },
}
`;
